src/main/scala/vsys/blockchain/contract/DataType.scala

```diff
@@ -1,34 +1,162 @@
 package vsys.blockchain.contract
 
 import vsys.blockchain.transaction.TransactionParser.{AmountLength, KeyLength, TimestampLength}
-
+import play.api.libs.json.{JsValue, Json}
+import com.google.common.primitives.{Ints, Longs, Shorts}
+import vsys.account.{AddressScheme, PublicKeyAccount}
+import scorex.crypto.encode.Base58
+import vsys.account.ContractAccount.ChecksumLength
+import vsys.blockchain.state.ByteStr
+import vsys.utils.crypto.hash.SecureCryptographicHash._
 import scala.util.Try
 
 object DataType extends Enumeration {
 
   val MaxShortTextLength = 140
   val MaxShortBytesLength = 255
   val MaxOpcBlockLength = 255
+  val MaxBigIntLength = 255       // less than 1024 bits
+
+  sealed case class DataTypeVal[T](dataType: Int, lenFixed: Boolean, maxLen: Int,
+                                   deserializer: Array[Byte] => T,
+                                   serializer: T => Array[Byte],
+                                   jsonifier: T => JsValue,
+                                   private val extValidator: Array[Byte] => Boolean) extends Val(dataType) {
+
+    def *(n: Int): Int = n * dataType
+
+    val jsonifierB: Array[Byte] => JsValue = b => jsonifier(deserializer(b))
+
+    // check length and other limits
+    val validator: Array[Byte] => Boolean = data => (((lenFixed && data.length == maxLen)
+        || (!lenFixed && data.length > 2 && Shorts.fromByteArray(data.take(2)) == data.drop(2).length && data.drop(2).length <= maxLen))
```

Comment:
How about `data.length == 2` case? e.g. empty shortBytes? may add a related unit test for this boundary case.

Reply:
Is Array[Byte](0, 0) a valid data?

Reply:
how to represent a `shortBytes` or `shortText` with zero length? I think `Array[Byte](0, 0)` is a valid data for `shortBytes` and `shortText` but not for `BigInteger`(if BigInt(0) is not an empty array)

but we can set `shortBytes` and `shortText` a non-empty array. will it be too strict?

Reply:
ok, changed to `data.length >= 2`